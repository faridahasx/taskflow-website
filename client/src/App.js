// External Imports
import { Suspense, lazy, useEffect, useLayoutEffect } from "react";
import { BrowserRouter, Routes, Route, Navigate } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
// MUI Components
import { AdapterDayjs } from "@mui/x-date-pickers/AdapterDayjs";
import { LocalizationProvider } from "@mui/x-date-pickers";
// Custom hooks
import useNetworkStatus from "./hooks/useNetworkStatus";
// Components
import LoadingPage from "./components/Loading/LoadingPage";
import ErrorBoundary from "./components/Error/ErrorBoundary";
import Message from "./components/Alerts/Message";
import Offline from "./components/Alerts/Offline";
// Styles
import "./assets/globals.css";

// Lazily loaded components
const Home = lazy(() => import("./pages/Home"));
const LoginSuccessRedirect = lazy(() => import("./pages/LoginSuccessRedirect"));
const Stats = lazy(() => import("./pages/Stats"));
const LandingPage = lazy(() => import("./pages/LandingPage"));


function App() {
  // Redux dispatch function
  const dispatch = useDispatch();
  // Check for user authentication status from Redux store
  const isLogged = useSelector((state) => state.auth.isLogged);
  // Check online status using custom hook
  const isOnline = useNetworkStatus();

  //Adjusts the viewport height on window resize
  useEffect(() => {
    const setInnerHeight = () => {
      document.documentElement.style.setProperty(
        "--vh",
        `${window.innerHeight}px`
      );
    };
    // Set initial inner height and add event listener for resize
    setInnerHeight();
    window.addEventListener("resize", setInnerHeight);
    // Cleanup on component unmount
    return () => {
      window.removeEventListener("resize", setInnerHeight);
    };
  }, []);

  // Checks for first login and dispatches the appropriate action
  useLayoutEffect(() => {
    const getFirstLogin = () => {
      const firstLogin = localStorage.getItem("firstLogin");
      dispatch({ type: "IS_LOGGED", payload: firstLogin ? true : false });
    };
    // Dispatches action only if not logged in
    if (isLogged === null) getFirstLogin();
  }, [isLogged, dispatch]);

  return (
    <LocalizationProvider dateAdapter={AdapterDayjs}>
      <BrowserRouter>
        <Message />
        {!isOnline && <Offline />}
        <ErrorBoundary>
          <Suspense fallback={<LoadingPage />}>
            <Routes>
              <Route exact path="/" element={isLogged===false? <LandingPage/>: isLogged===true?<Home />:null}/>
              <Route exact path="/stats" element={isLogged===false ? <Navigate replace to="/" /> : <Stats />}/>

              <Route path="/redirect" element={<LoginSuccessRedirect />} />
              <Route path="*" element={<Navigate to="/" />} />
            </Routes>
          </Suspense>
        </ErrorBoundary>
      </BrowserRouter>
    </LocalizationProvider>
  );
}

export default App;
