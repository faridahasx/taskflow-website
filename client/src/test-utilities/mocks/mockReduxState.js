import createMockStore from "redux-mock-store";

const mockStore = createMockStore([]);

const mockInitialState = {
  auth: { isLogged: null },
  alert: { message: "" },
  categories: [],
};

const mockAuthState = {
  ...mockInitialState,
  auth: { isLogged: true },
};

export { mockAuthState, mockStore };

export default mockInitialState;
