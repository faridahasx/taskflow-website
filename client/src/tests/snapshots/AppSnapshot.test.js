import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import App from "../../App";
import { MockIntersectionObserver } from "../mocks/mockClasses";
import mockInitialState, { mockStore } from "../mocks/mockReduxState";

test("App snapshot", () => {
  global.IntersectionObserver = MockIntersectionObserver;
  const { asFragment } = render(
    <Provider store={mockStore(mockInitialState)}>
      <App />
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot("App");
});
