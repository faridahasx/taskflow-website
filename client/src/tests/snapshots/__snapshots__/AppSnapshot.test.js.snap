// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot: App 1`] = `
<DocumentFragment>
  <div>
    <div
      class="loading-page-indicator"
    >
      <span
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate css-eglki6-MuiLinearProgress-root"
        role="progressbar"
      >
        <span
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate css-5ir5xx-MuiLinearProgress-bar1"
        />
        <span
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar2Indeterminate css-1r8wrcl-MuiLinearProgress-bar2"
        />
      </span>
    </div>
  </div>
</DocumentFragment>
`;
